### ISSUE TITLE
Hash Join expects distinct join keys

### ISSUE THREAD

**Issue by TomaszGaweda** [hazelcast/hazelcast-jet#1326]
Hi, it's me again ;)

I've discovered a bug(?) in Jet, library assumes that enrichment hashtable (in our case, HashMap) will be 1 -> 1 mapping. If there are more than one item with given key, Jet raises an error. I know in docs you wrote about "many to one" join, but this behaviour makes users unable to rewrite old SQL queries to Jet

This is not a valid behaviour in databases. Normally, enrichment tables are Multimaps - for each key, there are possibly few values. If `hashTable.get(key)` returns more than 1 resutl, then database manually checks if given key from main table is equal to elements in set of entries for given key.

 It's a normal behaviour, that join produces more than one row for one input row from main table, so it's IMHO a bug that Jet - while using the same name - offers very limited usage of hash joins.

This simple repro produces exception:
```java

import com.hazelcast.jet.Jet;
import com.hazelcast.jet.pipeline.BatchStage;
import com.hazelcast.jet.pipeline.Pipeline;
import com.hazelcast.jet.pipeline.Sinks;
import com.hazelcast.jet.pipeline.Sources;
import io.vavr.Tuple2;
import lombok.Value;

import java.io.Serializable;
import java.util.List;

import static com.hazelcast.jet.pipeline.JoinClause.onKeys;

public class ReproJoins {

    public static void main(String[] args) {
        var jet = Jet.newJetInstance();

        var aMap = jet.<String, EntityA>getMap("a");
        var bMap = jet.<Long, EntityB>getMap("b");

        List<EntityA> as = List.of(
                new EntityA("1", "1", "1"),
                new EntityA("2", "2", "1"),
                new EntityA("3", "3", "3")
        );
        List<EntityB> bs = List.of(
                new EntityB(1L, "1", "1"),
                new EntityB(2L, "1", "9"),
                new EntityB(3L, "1", "9"),
                new EntityB(4L, "3", "3")
        );

        // in DBs like Oracle, PostgreSQL: as /*hash*/ join bs on a.key = b.value1 produces: 4 rows

        for (EntityA a : as) {
            aMap.put(a.key, a);
        }
        for (EntityB b : bs) {
            bMap.put(b.key, b);
        }

        var pipeline = Pipeline.create();
        BatchStage<EntityB> readBStage = pipeline.drawFrom(Sources.map(bMap))
                .map(e -> e.getValue());

        pipeline.drawFrom(Sources.map(aMap))
                .map(e -> e.getValue())
                .hashJoin(readBStage, onKeys(EntityA::getKey, EntityB::getValue1), Tuple2::new)
                .drainTo(Sinks.logger());

        jet.newJob(pipeline).join();
    }

    @Value
    public static class EntityA implements Serializable  {
        String key;
        String value1;
        String value2;
    }
    @Value
    public static class EntityB  implements Serializable {
        Long key;
        String value1;
        String value2;
    }
}
```

Exception is: 
```
Duplicate values for key '1': 'ReproJoins.EntityB(key=2, value1=1, value2=9)' and 'ReproJoins.EntityB(key=3, value1=1, value2=9)'
```

While script:
```sql
create table EntityA (
  key varchar2(10),
  value1  varchar2(10),
  value2  varchar2(10)
);

create table EntityB (
   key number,
   value1  varchar2(10),
   value2  varchar2(10)
);

insert into EntityA(key, value1, value2)
values ('1', '1', '1');
insert into EntityA(key, value1, value2)
values ('2', '2', '2');
insert into EntityA(key, value1, value2)
values ('3', '3', '3');

insert into EntityB(key, value1, value2)
values (1, '1', '1');
insert into EntityB(key, value1, value2)
values (2, '1', '2');
insert into EntityB(key, value1, value2)
values (3, '1', '3');
insert into EntityB(key, value1, value2)
values (4, '3', '3');

commit;

select /*+ use_hash(e,b) */
a.key, a.value1, b.key, b.value1
from EntityA a
join EntityB b on a.key = b.value1;
```

that indeed uses hashjoin:
```
------------------------------------------------------------------------------
| Id  | Operation          | Name    | Rows  | Bytes | Cost (%CPU)| Time     |
------------------------------------------------------------------------------
|   0 | SELECT STATEMENT   |         |     4 |   136 |     5  (20)| 00:00:01 |
|*  1 |  HASH JOIN         |         |     4 |   136 |     5  (20)| 00:00:01 |
|   2 |   TABLE ACCESS FULL| ENTITYA |     3 |    42 |     2   (0)| 00:00:01 |
|   3 |   TABLE ACCESS FULL| ENTITYB |     4 |    80 |     2   (0)| 00:00:01 |
------------------------------------------------------------------------------
```

returns:

```
key value1 key value1
1	1	1	1
1	1	2	1
1	1	3	1
3	3	4	3
```

**Comment by viliam-durina** [comment]
This proposal looks reasonable, we might add it to the 3.1 release. The current release is already in feature-freeze state.

**Comment by TomaszGaweda** [comment]
@viliam-durina Ok, so if it's not rejected at all, we in our company will work on local fork, then switch back to mainline. Thanks, I will try to make all corrections as soon as possible

**Comment by TomaszGaweda** [comment]
Closed by #1327
